### DISCUSSION TITLE
13. HTTP에 대해 설명하고 특징을 설명해 주세요.

### DISCUSSION THREAD

**Discussion by eun-byeol** [jewelry-box/cs-interview#17] [General]
HTTP에 대해 설명하고 특징을 설명해 주세요.

**Comment by jongmee** [comment]
- HTTP는 Hyper Text Transfer Protocol로 클라이언트와 서버가 텍스트 기반 데이터를 주고 받을 때 사용하는 프로토콜이다.
- 네트워크 7계층에서 어플리케이션 계층에 해당한다.
- 무상태성을 가지고 요청-응답 구조이다.

**Comment by cutehumanS2** [comment]
HTTP는 응용 계층에서 정보를 주고받는 데 사용되는 프로토콜로, 클라이언트-서버 구조를 기반으로 요청과 응답을 주고받으며 동작합니다.  

HTTP의 주요 특징으로는 다음 네 가지가 있습니다.
1. **요청-응답 기반 프로토콜**
  클라이언트가 서버가 서로 HTTP 요청 메시지와 HTTP 응답 메시지를 주고받는 구조로 동작합니다.
    
2. **미디어 독립적 프로토콜**
  데이터의 미디어 타입(MIME 타입)과 관계없이 전송할 수 있습니다.
    
3. **무상태성(Stateless) 프로토콜**
    서버는 클라이언트의 이전 요청 정보를 기억하지 않으며, 매 요청을 독립적으로 처리합니다. 
    
4. **지속 연결(Persistent Connection) 프로토콜**
    HTTP/1.1부터 keep-alive를 기본으로 지원하여, 하나의 TCP 연결을 통해 여러 요청-응답을 지속적으로 처리할 수 있습니다.


<br>

**[관련 질문]**
[HTTP 무상태성은 무엇인가요? ‘오늘 하루 그만보기’ 같은 기능은 어떻게 구현하나요?](https://github.com/jewelry-box/cs-interview/discussions/18)

[참고한 자료]
<혼자 공부하는 네트워크> - 05-2. HTTP 중 “HTTP의 특성”

**Comment by cutehumanS2** [comment]
`추가 질문`

### HTTP가 비연결형(Connectionless)이라는 것은 무슨 뜻인가요?
- HTTP는 기본적으로 요청당 하나의 TCP 연결을 설정하고, 요청 완료 후 연결을 해제하기 때문에, 비연결형이라고 합니다.
- HTTP/1.1부터는 지속 연결(keep-alive) 기능이 도입되어, 하나의 TCP 연결로 여러 요청을 처리하여 성능을 개선하였습니다.

<hr>

### HTTP가 비연결형으로 설계된 이유는 무엇인가요?
- 네트워크 자원의 효율성을 확보하기 위함입니다.
- 요청-응답 후 연결을 종료하여 서버가 불필요한 연결을 유지하지 않아도 되므로, 동시 접속자 수를 늘리고 확장성을 높일 수 있습니다.

<hr>

### HTTP가 무상태성 프로토콜로 설계된 이유와, 이를 통해 얻을 수 있는 이점은 무엇인가요?
- **확장성 증가**
    - 서버가 클라이언트의 상태 정보를 유지하지 않으므로, 새로운 서버를 추가하거나 트래픽 증가에 대응하기 용이합니다.
    - 로드밸런싱 환경에서도 클라이언트 요청을 특정 서버에 종속시키지 않고, 분산 처리할 수 있습니다.
- **성능 향상 및 리소스 절약**
    - 모든 요청이 독립적이므로, 불필요한 상태 정보 저장에 필요한 리소스를 절약할 수 있습니다.
    - 따라서 메모리 사용량이 감소하고, 더 많은 요청을 동시에 처리할 수 있어 성능이 향상됩니다.
- **견고성 확보(문제 발생 시 대처 용이)**
    - 특정 클라이언트가 특정 서버에 종속되지 않기 때문에, 특정 서버에 문제가 생겨도 쉽게 다른 서버로 대체 가능합니다.
